cpp/merkletree/sparse_merkle_tree_test.cc

```diff
@@ -0,0 +1,315 @@
+      return tree_hasher_.HashLeaf(lo->second);
+    }
+    if (lo == hi) {
+      return HStarEmpty(n);
+    }
+
+    // DIFF: use BIGNUM, 'cos we'll get to values of O(1 << 256) here (!)
+    BIGNUM* split(BN_new());
+    CHECK_EQ(1, BN_set_word(split, 1));
+    CHECK_EQ(1, BN_lshift(split, split, n - 1));
+    CHECK_EQ(1, BN_add(split, split, offset));
+
+    auto i(lower_bound(lo, hi, split, KeyLess()));
+    const string ret(
+        tree_hasher_.HashChildren(HStar2b(n - 1, values, lo, i, offset),
+                                  HStar2b(n - 1, values, i, hi, split)));
+    BN_free(split);
+    return ret;
+  }
+
+  TreeHasher tree_hasher_;
+  vector<string> hStarEmptyCache_;
+};
+
+
+class SparseMerkleTreeTest : public testing::Test {
+ public:
+  SparseMerkleTreeTest()
+      : tree_hasher_(new Sha256Hasher),
+        tree_(new Sha256Hasher()),
+        rand_({1234}),
+        null_hashes_(CalculateNullHashes(tree_hasher_)) {
+  }
+
+ protected:
+  // Returns a Path with the high 64 bits set to |high|
+  Path PathHigh(uint64_t high) {
+    Path ret;
+    for (int i(ret.size() - sizeof(high) * 8); i < ret.size(); ++i) {
+      ret[i] = high & 0x01;
+      high >>= 1;
+    }
+    return ret;
+  }
+
+  // Returns a random Path.
+  Path RandomPath() {
+    Path ret;
+    for (int i(0); i < ret.size(); ++i) {
+      ret[i] = rand_() & 0x01;
+    }
+    return ret;
+  }
+
+  TreeHasher tree_hasher_;
+  SparseMerkleTree tree_;
+  mt19937 rand_;
+  const vector<string> null_hashes_;
+};
+
+
+TEST_F(SparseMerkleTreeTest, ReferenceEmptyTreeRootKAT) {
+  Reference ref(new Sha256Hasher);
+  EXPECT_EQ(kEmptyRootHashB64, ToBase64(ref.HStar2(256, ValueList())));
+}
+
+
+TEST_F(SparseMerkleTreeTest, EmptyTreeRootKAT) {
+  tree_.CurrentRoot();
+  EXPECT_EQ(kEmptyRootHashB64, ToBase64(tree_.CurrentRoot()));
+}
+
+
+TEST_F(SparseMerkleTreeTest, SimpleTest) {
+  Reference ref(new Sha256Hasher);
+  ValueList values;
+  for (auto r : vector<uint64_t>{0, 10}) {
+    const string value(to_string(r));
+    values.emplace_back(std::move(Value(r, value)));
+    const Path p(r);
+    tree_.SetLeaf(p, value);
+    tree_.CurrentRoot();
+  }
+  EXPECT_EQ(ToBase64(ref.HStar2(256, values)), ToBase64(tree_.CurrentRoot()));
+}
+
+
+TEST_F(SparseMerkleTreeTest, RandomReferenceTest) {
+  Reference ref(new Sha256Hasher);
+  ValueList values;
+  LOG(INFO) << "Setup";
+  for (int i(0); i < 1000000; ++i) {
+    uint64_t r(rand_() + i);
+    const string value(to_string(r));
+    values.emplace_back(std::move(Value(r, value)));
+    const Path p(r);
+    tree_.SetLeaf(p, value);
+  }
+  LOG(INFO) << "Calculating SMT Root";
+  const string smt_root(tree_.CurrentRoot());
+  LOG(INFO) << "Calculating Reference Root";
+  const string ref_root(ref.HStar2(256, values));
+  LOG(INFO) << "Comparing";
+  EXPECT_EQ(ToBase64(ref_root), ToBase64(smt_root));
+}
+
+
+/*
+These are mainly useful to measuring memory consumption, but they take a while
```

Comment:
Use `DISABLED_`, so they're still compiled. And if you're going to disable a big block of code like that, always use `#if 0` anyway, to protect against comments in the block!